src/lib.rs

```diff
@@ -174,15 +174,25 @@ fn compress<T: Clone + Copy + PartialEq>(
     let mut r = Vec::new(); // Result vector
     r.resize(row_length, empty_val);
 
-    let mut dv = Vec::new(); // displacement vector
-    dv.resize(sorted.len(), 0);
+    let mut dv = vec![0; sorted.len()]; // displacement vector
 
+    let mut tmp = Vec::new();
     for s in sorted {
-        let slice = &vec[s * row_length..(s + 1) * row_length];
+        // The row we're about to iterate over typically contains mostly empty values that can
+        // never succeed with `fits`. We pre-filter out all those empty values up-front, such that
+        // `tmp` contains `(index, non-empty-value)` pairs that we can then pass to `fits`. Because
+        // this is such a tight loop, we reuse the same `Vec` to avoid repeated allocations.
+        tmp.clear();
+        for (i, v) in vec[s * row_length..(s + 1) * row_length].iter().enumerate() {
```

Comment:
This is a fairly minor stylistic choice so by all means disregard if it doesn't suit your fancy...
anyhow, I wonder about instead of this for loop, something like

```
tmp.extend(vec[s * row_length..(s + 1) * row_length].iter().enumerate().filter(|v| v != empty_val));
```

Reply:
Good idea. Fixed in https://github.com/softdevteam/sparsevec/pull/25/commits/2d37f94531570ad8458204338566ce14aceff57e. Will need squashing.